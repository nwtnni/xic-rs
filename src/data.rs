pub mod token;
pub mod ast;
pub mod typ;
pub mod operand;
