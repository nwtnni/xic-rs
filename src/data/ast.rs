use std::cell::Cell;
use std::fmt;

use crate::data::sexp::Serialize as _;
use crate::data::span::Span;
use crate::data::symbol::Symbol;

/// Represents a Xi interface file.
#[derive(Clone, Debug)]
pub struct Interface {
    pub uses: Vec<Use>,
    pub items: Vec<ItemSignature>,
}

impl fmt::Display for Interface {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a Xi source file.
#[derive(Clone, Debug)]
pub struct Program {
    pub uses: Vec<Use>,
    pub items: Vec<Item>,
}

impl fmt::Display for Program {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a use statement for importing interfaces.
#[derive(Clone, Debug)]
pub struct Use {
    pub name: Symbol,
    pub span: Span,
}

impl fmt::Display for Use {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum ItemSignature {
    Class(ClassSignature),
    Function(FunctionSignature),
}

impl fmt::Display for ItemSignature {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Item {
    Global(Global),
    Class(Class),
    Function(Function),
}

impl fmt::Display for Item {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Global {
    Declaration(Declaration),
    Initialization(Initialization),
}

impl fmt::Display for Global {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct Initialization {
    pub declarations: Vec<Option<SingleDeclaration>>,
    pub expression: Expression,
    pub span: Span,
}

impl fmt::Display for Initialization {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct ClassSignature {
    pub name: Symbol,
    pub extends: Option<Symbol>,
    pub methods: Vec<FunctionSignature>,
    pub span: Span,
}

impl fmt::Display for ClassSignature {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct Class {
    pub name: Symbol,
    pub extends: Option<Symbol>,
    pub items: Vec<ClassItem>,
    pub span: Span,
}

impl fmt::Display for Class {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum ClassItem {
    Field(Declaration),
    Method(Function),
}

impl fmt::Display for ClassItem {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

pub trait Callable {
    fn name(&self) -> Symbol;
    fn parameters(&self) -> &[SingleDeclaration];
    fn returns(&self) -> &[Type];
}

macro_rules! impl_callable {
    ($type:ty) => {
        impl Callable for $type {
            fn name(&self) -> Symbol {
                self.name
            }

            fn parameters(&self) -> &[SingleDeclaration] {
                &self.parameters
            }

            fn returns(&self) -> &[Type] {
                &self.returns
            }
        }
    };
}

/// Represents a function signature (i.e. without implementation).
#[derive(Clone, Debug)]
pub struct FunctionSignature {
    pub name: Symbol,
    pub parameters: Vec<SingleDeclaration>,
    pub returns: Vec<Type>,
    pub span: Span,
}

impl_callable!(FunctionSignature);

impl fmt::Display for FunctionSignature {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a function definition (i.e. with implementation).
#[derive(Clone, Debug)]
pub struct Function {
    pub name: Symbol,
    pub parameters: Vec<SingleDeclaration>,
    pub returns: Vec<Type>,
    pub statements: Statement,
    pub span: Span,
}

impl_callable!(Function);

impl fmt::Display for Function {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a primitive type.
#[derive(Clone, Debug)]
pub enum Type {
    Bool(Span),
    Int(Span),
    Class(Symbol, Span),
    Array(Box<Type>, Option<Expression>, Span),
}

impl Type {
    pub fn has_length(&self) -> bool {
        match self {
            Type::Bool(_) | Type::Int(_) | Type::Class(_, _) => false,
            Type::Array(r#type, length, _) => length.is_some() || r#type.has_length(),
        }
    }

    pub fn span(&self) -> Span {
        match self {
            Type::Bool(span) | Type::Int(span) | Type::Class(_, span) | Type::Array(_, _, span) => {
                *span
            }
        }
    }
}

impl PartialEq for Type {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Type::Bool(_), Type::Bool(_)) | (Type::Int(_), Type::Int(_)) => true,
            (Type::Array(lhs, _, _), Type::Array(rhs, _, _)) => lhs == rhs,
            _ => false,
        }
    }
}

impl Eq for Type {}

impl fmt::Display for Type {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents an imperative statement.
#[derive(Clone, Debug)]
pub enum Statement {
    /// Assignment
    Assignment(Expression, Expression, Span),

    /// Procedure call
    Call(Call),

    /// Initialization
    Initialization(Initialization),

    /// Variable declaration
    Declaration(Declaration, Span),

    /// Return statement
    Return(Vec<Expression>, Span),

    /// Statement block
    Sequence(Vec<Statement>, Span),

    /// If-else block
    If(Expression, Box<Statement>, Option<Box<Statement>>, Span),

    /// While block
    While(Do, Expression, Box<Statement>, Span),

    /// Break statement
    Break(Span),
}

impl Statement {
    pub fn span(&self) -> Span {
        match self {
            Statement::Call(call) => call.span,
            Statement::Initialization(initialization) => initialization.span,
            Statement::Assignment(_, _, span)
            | Statement::Declaration(_, span)
            | Statement::Return(_, span)
            | Statement::Sequence(_, span)
            | Statement::If(_, _, _, span)
            | Statement::While(_, _, _, span)
            | Statement::Break(span) => *span,
        }
    }
}

impl fmt::Display for Statement {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Do {
    Yes,
    No,
}

/// Represents an expression (i.e. a term that can be evaluated).
#[derive(Clone, Debug)]
pub enum Expression {
    /// Boolean literal
    Boolean(bool, Span),

    /// Char literal
    Character(char, Span),

    /// String literal
    String(String, Span),

    /// Integer literal
    Integer(i64, Span),

    /// Null literal
    Null(Span),

    /// Class reference
    This(Span),

    /// Variable
    Variable(Symbol, Span),

    /// Array literal
    Array(Vec<Expression>, Span),

    /// Binary operation
    Binary(Cell<Binary>, Box<Expression>, Box<Expression>, Span),

    /// Unary operation
    Unary(Unary, Box<Expression>, Span),

    /// Array index
    Index(Box<Expression>, Box<Expression>, Span),

    /// Array length
    Length(Box<Expression>, Span),

    /// Function call
    Call(Call),

    /// Dot operator
    Dot(Box<Expression>, Symbol, Span),

    /// Class constructor
    New(Symbol, Span),
}

impl Expression {
    pub fn span(&self) -> Span {
        match self {
            Expression::Boolean(_, span)
            | Expression::Character(_, span)
            | Expression::String(_, span)
            | Expression::Integer(_, span)
            | Expression::Null(span)
            | Expression::This(span)
            | Expression::Variable(_, span)
            | Expression::Array(_, span)
            | Expression::Binary(_, _, _, span)
            | Expression::Unary(_, _, span)
            | Expression::Index(_, _, span)
            | Expression::Length(_, span)
            | Expression::Dot(_, _, span)
            | Expression::New(_, span) => *span,
            Expression::Call(call) => call.span,
        }
    }

    pub fn span_mut(&mut self) -> &mut Span {
        match self {
            Expression::Boolean(_, span)
            | Expression::Character(_, span)
            | Expression::String(_, span)
            | Expression::Integer(_, span)
            | Expression::Null(span)
            | Expression::This(span)
            | Expression::Variable(_, span)
            | Expression::Array(_, span)
            | Expression::Binary(_, _, _, span)
            | Expression::Unary(_, _, span)
            | Expression::Index(_, _, span)
            | Expression::Length(_, span)
            | Expression::Dot(_, _, span)
            | Expression::New(_, span) => span,
            Expression::Call(call) => &mut call.span,
        }
    }
}

impl fmt::Display for Expression {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a function call.
#[derive(Clone, Debug)]
pub struct Call {
    pub function: Box<Expression>,
    pub arguments: Vec<Expression>,
    pub span: Span,
}

impl fmt::Display for Call {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub enum Declaration {
    Multiple(MultipleDeclaration),
    Single(SingleDeclaration),
}

impl Declaration {
    pub fn has_length(&self) -> bool {
        match self {
            Declaration::Multiple(multiple) => multiple.has_length(),
            Declaration::Single(single) => single.has_length(),
        }
    }

    pub fn span(&self) -> Span {
        match self {
            Declaration::Multiple(multiple) => multiple.span(),
            Declaration::Single(single) => single.span(),
        }
    }
}

impl fmt::Display for Declaration {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct MultipleDeclaration {
    pub names: Vec<Symbol>,
    pub r#type: Type,
    pub span: Span,
}

impl MultipleDeclaration {
    pub fn new(names: Vec<Symbol>, r#type: Type, span: Span) -> Self {
        Self {
            names,
            r#type,
            span,
        }
    }

    pub fn has_length(&self) -> bool {
        self.r#type.has_length()
    }

    pub fn span(&self) -> Span {
        self.span
    }
}

impl fmt::Display for MultipleDeclaration {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct SingleDeclaration {
    pub name: Symbol,
    pub r#type: Type,
    pub span: Span,
}

impl SingleDeclaration {
    pub fn new(name: Symbol, r#type: Type, span: Span) -> Self {
        Self { name, r#type, span }
    }

    pub fn has_length(&self) -> bool {
        self.r#type.has_length()
    }

    pub fn span(&self) -> Span {
        self.span
    }
}

impl fmt::Display for SingleDeclaration {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a binary operator.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Binary {
    Mul,
    Hul,
    Div,
    Mod,
    Add,
    Cat,
    Sub,
    Lt,
    Le,
    Ge,
    Gt,
    Eq,
    Ne,
    And,
    Or,
}

impl fmt::Display for Binary {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

/// Represents a unary operator.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Unary {
    Neg,
    Not,
}

impl fmt::Display for Unary {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}
