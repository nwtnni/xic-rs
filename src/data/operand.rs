use std::sync::atomic::{AtomicUsize, Ordering};

use crate::data::symbol;

static LABELS: AtomicUsize = AtomicUsize::new(0);
static TEMPS: AtomicUsize = AtomicUsize::new(0);

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Immediate {
    Constant(i64),
    Label(Label),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Label {
    Fixed(symbol::Symbol),
    Fresh(symbol::Symbol, usize),
}

impl Label {
    pub fn fresh(label: &'static str) -> Self {
        let index = LABELS.fetch_add(1, Ordering::SeqCst);
        let symbol = symbol::intern_static(label);
        Label::Fresh(symbol, index)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Temporary {
    Register(Register),
    Argument(usize),
    Return(usize),
    Fresh(symbol::Symbol, usize),
}

impl Temporary {
    pub fn fresh(label: &'static str) -> Self {
        let index = TEMPS.fetch_add(1, Ordering::SeqCst);
        let symbol = symbol::intern_static(label);
        Temporary::Fresh(symbol, index)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Register {
    Rax,
    Rbx,
    Rcx,
    Rdx,
    Rbp,
    Rsp,
    Rsi,
    Rdi,
    R8,
    R9,
    R10,
    R11,
    R12,
    R13,
    R14,
    R15,
}

pub trait Operand: Copy + Eq + std::hash::Hash + std::fmt::Debug + PartialOrd + Ord {}
impl Operand for Temporary {}
impl Operand for Register {}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Memory<T: Operand> {
    R(T),
    RO(T, i32),
}
