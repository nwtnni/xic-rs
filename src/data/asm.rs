use crate::data::ir;
use crate::data::operand;

pub type Unit<T> = ir::Unit<Function<T>>;
pub type Function<T> = ir::Function<Vec<Assembly<T>>>;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Assembly<T> {
    Binary(Binary, operand::Binary<T>),
    Unary(Unary, operand::Unary<T>),
    Nullary(Nullary),
    Label(operand::Label),
    Directive(Directive),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Directive {
    Intel,
    Align(usize),
    Local(operand::Label),
    Global(operand::Label),
    Quad(Vec<i64>),
    Data,
    Text,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Binary {
    Add,
    Sub,
    And,
    Or,
    Xor,
    Cmp,
    Mov,
    Lea,
}

impl From<ir::Binary> for Binary {
    fn from(binary: ir::Binary) -> Self {
        match binary {
            ir::Binary::Add => Binary::Add,
            ir::Binary::Sub => Binary::Sub,
            ir::Binary::Xor => Binary::Xor,
            ir::Binary::And => Binary::And,
            ir::Binary::Or => Binary::Or,
            _ => panic!("[INTERNAL ERROR]: converting unsupported IR operator"),
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Unary {
    Neg,
    Push,
    Pop,
    Call { arguments: usize, returns: usize },
    Mul,
    Div(Division),
    Jmp,
    Jcc(Condition),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Nullary {
    Cqo,
    Ret,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Condition {
    L,
    Le,
    Ge,
    G,
    Ne,
    E,
}

impl From<ir::Condition> for Condition {
    fn from(condition: ir::Condition) -> Self {
        match condition {
            ir::Condition::Lt => Condition::L,
            ir::Condition::Le => Condition::Le,
            ir::Condition::Ge => Condition::Ge,
            ir::Condition::Gt => Condition::G,
            ir::Condition::Ne => Condition::Ne,
            ir::Condition::Eq => Condition::E,
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Division {
    Quotient,
    Remainder,
}
