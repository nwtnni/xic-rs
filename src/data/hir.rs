use crate::data::ir;
use crate::data::operand;
use crate::data::symbol;

#[derive(Clone, Debug)]
pub struct Function {
    pub name: symbol::Symbol,
    pub statements: Statement,
}

pub enum Tree {
    Expression(Expression),
    Condition(Condition),
}

pub type Condition = Box<dyn FnOnce(operand::Label, operand::Label) -> Statement>;

impl From<Condition> for Tree {
    fn from(condition: Condition) -> Self {
        Tree::Condition(condition)
    }
}

impl From<Tree> for Condition {
    fn from(tree: Tree) -> Self {
        match tree {
            Tree::Condition(condition) => condition,
            Tree::Expression(expression) => Box::new(move |r#true, r#false| {
                let condition = Expression::Binary(
                    ir::Binary::Eq,
                    Box::new(expression),
                    Box::new(Expression::Integer(1)),
                );
                Statement::CJump(condition, r#true, r#false)
            }),
        }
    }
}

#[macro_export]
macro_rules! hir {
    ((CONST $($integer:tt)+)) => {
        crate::data::hir::Expression::from(hir!($($integer)+))
    };
    ((NAME $($label:tt)+)) => {
        crate::data::hir::Expression::from(hir!($($label)+))
    };
    ((TEMP $($temporary:tt)+)) => {
        crate::data::hir::Expression::from(hir!($($temporary)+))
    };
    ((MEM $expression:tt)) => {
        crate::data::hir::Expression::Memory(Box::new(hir!($expression)))
    };
    ((CALL $function:tt $returns:tt $($argument:tt)*)) => {
        crate::data::hir::Expression::Call(
            Box::new(hir!($function)),
            vec![$(hir!($argument),)*],
            $returns,
        )
    };
    ((ESEQ $statement:tt $expression:tt)) => {
        crate::data::hir::Expression::Sequence(
            Box::new(hir!($statement)),
            Box::new(hir!($expression)),
        )
    };

    ((JUMP $label:ident)) => {
        crate::data::hir::Statement::Jump($label)
    };
    ((CJUMP $condition:tt $r#true:ident $r#false:ident)) => {
        crate::data::hir::Statement::CJump(
            hir!($condition),
            $r#true,
            $r#false,
        )
    };
    ((LABEL $label:tt)) => {
        crate::data::hir::Statement::Label($label)
    };
    ((EXP $expression:tt)) => {
        crate::data::hir::Statement::Expression(hir!($expression))
    };
    ((MOVE $into:tt $from:tt)) => {
        crate::data::hir::Statement::Move(
            hir!($into),
            hir!($from),
        )
    };
    ((RETURN)) => {
        crate::data::hir::Statement::Return
    };
    ((SEQ $statement:tt $($statements:tt)+)) => {
        crate::data::hir::Statement::Sequence(vec![
            hir!($statement),
            $(hir!($statements),)*
        ])
    };
    ((SEQ $statements:expr)) => {
        crate::data::hir::Statement::Sequence($statements)
    };

    (($binary:ident $left:tt $right:tt)) => {
        crate::data::hir::Expression::Binary(
            $binary,
            Box::new(hir!($left)),
            Box::new(hir!($right)),
        )
    };
    ($expression:expr) => {
        $expression
    }
}

#[derive(Clone, Debug)]
pub enum Expression {
    Integer(i64),
    Label(operand::Label),
    Temporary(operand::Temporary),
    Memory(Box<Expression>),
    Binary(ir::Binary, Box<Expression>, Box<Expression>),
    Call(Box<Expression>, Vec<Expression>, usize),
    Sequence(Box<Statement>, Box<Expression>),
}

impl From<operand::Temporary> for Expression {
    fn from(temporary: operand::Temporary) -> Self {
        Self::Temporary(temporary)
    }
}

impl From<operand::Label> for Expression {
    fn from(label: operand::Label) -> Self {
        Self::Label(label)
    }
}

impl From<i64> for Expression {
    fn from(integer: i64) -> Self {
        Self::Integer(integer)
    }
}

impl From<Expression> for Tree {
    fn from(expression: Expression) -> Self {
        Tree::Expression(expression)
    }
}

impl From<Tree> for Expression {
    fn from(tree: Tree) -> Self {
        match tree {
            Tree::Expression(expression) => expression,
            Tree::Condition(condition) => {
                let r#true = operand::Label::fresh("true");
                let r#false = operand::Label::fresh("false");
                let value = Expression::Temporary(operand::Temporary::fresh("bool"));

                let sequence = vec![
                    Statement::Move(value.clone(), Expression::Integer(0)),
                    condition(r#true, r#false),
                    Statement::Label(r#true),
                    Statement::Move(value.clone(), Expression::Integer(1)),
                    Statement::Label(r#false),
                ];

                Expression::Sequence(Box::new(Statement::Sequence(sequence)), Box::new(value))
            }
        }
    }
}

#[derive(Clone, Debug)]
pub enum Statement {
    Jump(operand::Label),
    CJump(Expression, operand::Label, operand::Label),
    Label(operand::Label),
    Expression(Expression),
    Move(Expression, Expression),
    Return,
    Sequence(Vec<Statement>),
}
