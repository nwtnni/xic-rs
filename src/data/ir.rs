use std::collections::HashMap;

use crate::data::hir;
use crate::data::lir;
use crate::data::operand;
use crate::util::symbol;

#[derive(Clone, Debug)]
pub struct Unit<F: IR> {
    pub name: symbol::Symbol,
    pub funs: HashMap<symbol::Symbol, F>,
    pub data: HashMap<symbol::Symbol, operand::Label>,
}

pub trait IR {}
impl IR for hir::Fun {}
impl IR for lir::Fun {}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Bin {
    Add,
    Sub,
    Mul,
    Hul,
    Div,
    Mod,
    Xor,
    Ls,
    Rs,
    ARs,
    Lt,
    Le,
    Ge,
    Gt,
    Ne,
    Eq,
    And,
    Or,
}
