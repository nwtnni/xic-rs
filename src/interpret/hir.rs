#![allow(dead_code)]

use std::io;

use anyhow::anyhow;
use anyhow::Context as _;

use crate::data::hir;
use crate::data::ir;
use crate::data::operand;
use crate::data::sexp::Serialize as _;
use crate::data::symbol;
use crate::interpret::postorder;
use crate::interpret::Global;
use crate::interpret::Local;
use crate::interpret::Operand;
use crate::interpret::Postorder;
use crate::interpret::Value;

pub fn interpret_hir<'io, R: io::BufRead + 'io, W: io::Write + 'io>(
    unit: &ir::Unit<hir::Function>,
    stdin: R,
    stdout: W,
) -> anyhow::Result<()> {
    let unit = Postorder::traverse_hir_unit(unit);

    let mut global = Global::new(&unit.data, stdin, stdout);
    let mut local = Local::new(
        &unit,
        &symbol::intern_static("_Imain_paai"),
        &[Value::Integer(0)],
    );

    assert!(local.interpret_hir(&unit, &mut global)?.is_empty());

    Ok(())
}

impl<'a> Local<'a, postorder::Hir<'a>> {
    fn interpret_hir(
        &mut self,
        unit: &ir::Unit<Postorder<postorder::Hir<'a>>>,
        global: &mut Global,
    ) -> anyhow::Result<Vec<Value>> {
        loop {
            let instruction = match self.step() {
                Some(instruction) => instruction,
                None => return Ok(Vec::new()),
            };

            match instruction {
                postorder::Hir::Expression(expression) => {
                    self.interpret_expression(unit, global, expression)?
                }
                postorder::Hir::Statement(statement) => {
                    if let Some(r#returns) = self.interpret_statement(global, statement)? {
                        return Ok(r#returns);
                    }
                }
            }
        }
    }

    fn interpret_expression(
        &mut self,
        unit: &ir::Unit<Postorder<postorder::Hir<'a>>>,
        global: &mut Global,
        expression: &hir::Expression,
    ) -> anyhow::Result<()> {
        log::trace!("E> {}", expression.sexp());
        match expression {
            hir::Expression::Sequence(_, _) => unreachable!(),
            hir::Expression::Integer(integer) => self.push(Operand::Integer(*integer)),
            hir::Expression::Label(label) => self.push(Operand::Label(*label, 8)),
            hir::Expression::Temporary(temporary) => self.push(Operand::Temporary(*temporary)),
            hir::Expression::Memory(_) => {
                let address = self.pop(global);
                self.push(Operand::Memory(address));
            }
            hir::Expression::Binary(binary, _, _) => self.interpret_binary(global, binary),
            hir::Expression::Call(_, arguments) => {
                let arguments = self.pop_arguments(global, arguments.len());
                let name = self.pop_name(global);

                log::info!("Calling function {} with arguments {:?}", name, arguments);

                let r#returns = global
                    .interpret_library(name, &arguments)
                    .unwrap_or_else(|| {
                        Local::new(unit, &name, &arguments).interpret_hir(unit, global)
                    })
                    .with_context(|| anyhow!("Calling function {}", name))?;

                if !r#returns.is_empty() {
                    self.push(r#returns[0].into_operand());
                }

                r#returns
                    .into_iter()
                    .enumerate()
                    .for_each(|(index, r#return)| {
                        self.insert(operand::Temporary::Return(index), r#return)
                    });
            }
        }

        Ok(())
    }

    fn interpret_statement(
        &mut self,
        global: &mut Global,
        statement: &hir::Statement,
    ) -> anyhow::Result<Option<Vec<Value>>> {
        log::trace!("S> {}", statement.sexp());
        match statement {
            hir::Statement::Expression(_) => unreachable!(),
            hir::Statement::Label(_) => unreachable!(),
            hir::Statement::Sequence(_) => unreachable!(),
            hir::Statement::Jump(_) => {
                self.interpret_jump(global);
                return Ok(None);
            }
            hir::Statement::CJump(_, r#true, r#false) => {
                self.interpret_cjump(global, r#true, r#false);
                return Ok(None);
            }
            hir::Statement::Move(_, _) => self.interpret_move(global),
            hir::Statement::Return(r#returns) => {
                return Ok(Some(self.pop_arguments(global, r#returns.len())));
            }
        }

        Ok(None)
    }
}
