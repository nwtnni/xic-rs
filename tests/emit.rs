use std::io;
use std::path::Path;

use xic::data::hir;
use xic::data::ir;

fn compile(path: &str) -> ir::Unit<hir::Function> {
    let tokens = xic::api::lex(Path::new(path)).unwrap();
    let program = xic::api::parse(tokens).unwrap();
    let context = xic::api::check(Path::new(path).parent().unwrap(), &program).unwrap();
    xic::api::emit_hir(Path::new(path), &program, &context)
}

#[test_generator::test_resources("tests/execute/*.xi")]
pub fn interpret(path: &str) {
    let hir = compile(path);
    let lir = xic::api::emit_lir(&hir);

    let mut hir_stdin = io::Cursor::new(Vec::new());
    let mut hir_stdout = io::Cursor::new(Vec::new());
    xic::api::interpret_hir(&hir, &mut hir_stdin, &mut hir_stdout).unwrap();

    let mut lir_stdin = io::Cursor::new(Vec::new());
    let mut lir_stdout = io::Cursor::new(Vec::new());
    xic::api::interpret_lir(&lir, &mut lir_stdin, &mut lir_stdout).unwrap();

    let hir_stdout = String::from_utf8(hir_stdout.into_inner()).unwrap();
    let lir_stdout = String::from_utf8(lir_stdout.into_inner()).unwrap();

    pretty_assertions::assert_eq!(hir_stdout, lir_stdout);
    insta::assert_snapshot!(path, lir_stdout);
}

#[test_generator::test_resources("tests/execute/*.xi")]
pub fn reorder(path: &str) {
    let hir = compile(path);
    let lir = xic::api::emit_lir(&hir);

    let mut lir_stdin = io::Cursor::new(Vec::new());
    let mut lir_stdout = io::Cursor::new(Vec::new());
    xic::api::interpret_lir(&lir, &mut lir_stdin, &mut lir_stdout).unwrap();

    let cfg = xic::api::construct_control_flow(&lir);
    let cfg = xic::api::destruct_control_flow(&cfg);

    let mut cfg_stdin = io::Cursor::new(Vec::new());
    let mut cfg_stdout = io::Cursor::new(Vec::new());
    xic::api::interpret_lir(&cfg, &mut cfg_stdin, &mut cfg_stdout).unwrap();

    let lir_stdout = String::from_utf8(lir_stdout.into_inner()).unwrap();
    let cfg_stdout = String::from_utf8(cfg_stdout.into_inner()).unwrap();

    pretty_assertions::assert_eq!(lir_stdout, cfg_stdout);
}
