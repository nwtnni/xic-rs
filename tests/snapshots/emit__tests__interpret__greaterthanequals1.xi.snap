---
source: tests/emit.rs
assertion_line: 41
expression: lir_stdout
---
2 >= 2 = True
